From 7e51346ad672343d3fcb03f894265c6fc10a02a8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Rapha=C3=ABl=20Hertzog?= <raphael@offensive-security.com>
Date: Sat, 18 Jan 2020 22:37:55 +0100
Subject: [PATCH 1/2] Fix resolution of virtual packages by reprepro mirror
 tool

There are multiple issues that had to be fixed:
- the path of the Packages files was wrong
- the code assumed the files were compressed, they are not
- the code did not work when multiple files matched (i.e. with multiple
  components) because grep would prepend the filename and a colon
  thus breaking "cud -d: -f2"
- the code did not work with versioned provides, we are now parsing
  the dependency correctly but we are ignoring the version constraint
- the number of provides in Debian grew very fast lately due to rust
  packages and the "providescheck" helper script failed with "Argument
  list too long", it has been rewritten to use file-backed package lists
  as input
- grep-dctrl exits with 1 and breaks the script when there's no result
  for the query, so calls need to be protected with "|| true"

Closes: #949255
---
 providecheck          |  7 +++++--
 tools/mirror/reprepro | 22 +++++++++++++++++-----
 2 files changed, 22 insertions(+), 7 deletions(-)

diff --git a/providecheck b/providecheck
index fad6f82..5a8c823 100755
--- a/providecheck
+++ b/providecheck
@@ -4,8 +4,11 @@
 
 import sys
 
-provides=sys.argv[1].split()
-packages=sys.argv[2].split()
+with open(sys.argv[1]) as fd_provides:
+    provides = [ x.strip() for x in fd_provides.readlines() ]
+
+with open(sys.argv[2]) as fd_packages:
+    packages = [ x.strip() for x in fd_packages.readlines() ]
 
 for package in packages:
     if provides.count(package) > 0:
diff --git a/tools/mirror/reprepro b/tools/mirror/reprepro
index 6e0da49..366e77e 100644
--- a/tools/mirror/reprepro
+++ b/tools/mirror/reprepro
@@ -6,12 +6,16 @@ package_lists=""
 for a in $ARCHES ; do
     for component in $mirror_components ; do
         package_lists="$package_lists $MIRROR/dists/$CODENAME/$component/binary-$a/Packages"
-        upstream_package_lists="$upstream_package_lists $MIRROR/lists/${CODENAME}_default_deb_${component}_${a}"
+        upstream_package_lists="$upstream_package_lists $MIRROR/lists/default_${CODENAME}_${component}_${a}_Packages"
     done
 done
 
+mytmpdir=$(mktemp -d)
+trap "rm -rf $mytmpdir" EXIT
+
 # get a list of provides
-provides="$(zgrep ^Provides $upstream_package_lists | cut -d : -f 2 | tr ',' '\n' | sort -u)"
+grep --no-filename ^Provides $upstream_package_lists | cut -d : -f 2- | tr ',' '\n' | perl -np -e 's/\s+\(.*\)//; s/^\s*//' | sort -u >$mytmpdir/provides
+
 for i in $(seq $reprepro_retries); do
     cp $conf/package-list $conf/package-list.old
 
@@ -22,14 +26,22 @@ for i in $(seq $reprepro_retries); do
         recommends=""
     fi
 
-    provide_dependencies="$(providecheck "$provides" "$all_packages $dependencies" | sort -u)"
+    for p in $all_packages $dependencies ; do
+        echo $p
+    done > $mytmpdir/packages
+    provide_dependencies=$(providecheck "$mytmpdir/provides" "$mytmpdir/packages" | sort -u)
     if [ -n "$provide_dependencies" ]; then
         for p in $provide_dependencies ; do
             # see if a package already satisfies the provides
-            satisfies="$(grep-dctrl -n --show-field Package --field Provides --field Package $p $package_lists)"
+            satisfies="$(grep-dctrl -n --show-field Package --field Provides --field Package $p $package_lists || true)"
             if [ -z "$satisfies" ]; then
                 # search for other packages that provide the desired features
-                dependencies="$dependencies $(zcat $upstream_package_lists | grep-dctrl -n --show-field Package --field Provides $p)"
+                providers=$(grep-dctrl -n --show-field Package --field Provides $p $upstream_package_lists || true)
+                if [ -z "$providers" ]; then
+                    echo "failed to find a package providing the virtual package '$p'"
+                    exit 1
+                fi
+                dependencies="$dependencies $providers"
             fi
         done
     fi
-- 
2.29.2

